### PR TITLE
Use retainedSize instead of estimateFlatSize in LocalExchange

### PR DESCRIPTION
**Pull request by Yuhta**
Summary:
We observed some OOM happening because `LocalExchange` is buffering the
input vectors more than the memory limit.  These vectors contain dictionary with
very few indices but a large alphabet, i.e. result of filtering.  In
`LocalExchange` we use `estimateFlatSize` to calculate the memory usage of
buffered data, which only counts the rows selected, so the size is heavily
underestimated.  We fix this by using `retainedSize` instead of
`estimateFlatSize` in `LocalExchange` to count the memory usage more properly.

Differential Revision: D59336200

### CODE REVIEW

File: velox/exec/LocalPartition.cpp [hunk 1]
```diff
@@ -77,19 +77,18 @@ void LocalExchangeQueue::noMoreProducers() {
 
 BlockingReason LocalExchangeQueue::enqueue(
     RowVectorPtr input,
+    int64_t size,
```
Comment (mbasmanova) [inline velox/exec/LocalPartition.cpp:80]: nit: perhaps, rename to inputBytes or similar to make it clear that this is size in bytes, not number of rows

**Review by mbasmanova — APPROVED** [review]
@Yuhta Looks good to me. I wonder if we still run into a problem where we queue too little data and produce lots of tiny vectors (assuming each small vector gets partitioned N ways).

**Comment by Yuhta** [comment]
@mbasmanova The vector fragmentation within `LocalExchange` still exists, I do not change that part in this PR.  To solve that we need to merge the output from queue.

**Review by xiaoxmeng — APPROVED** [review]
@Yuhta thanks!

**Pull request closed without merging**